gnucash/gnucash-cli.cpp

```diff
@@ -126,11 +144,117 @@ may be specified to describe some saved options.\n"
 
 }
 
+struct cli_struct
+{
+    boost::optional<std::string>& script;
+    bool interactive;
+    bool verbose;
+};
+
+static void
+run_guile_cli (void *data, [[maybe_unused]] int argc, [[maybe_unused]] char **argv)
+{
+    auto args = static_cast<cli_struct*>(data);
+    scm_c_use_module ("system repl repl");
+    scm_c_use_module ("gnucash core-utils");
+    scm_c_use_module ("gnucash engine");
+    scm_c_use_module ("gnucash app-utils");
+    scm_c_use_module ("gnucash report");
+    scm_c_use_module ("ice-9 readline");
+    scm_c_eval_string ("(activate-readline)");
+    if (args->script)
+    {
+        if (args->verbose)
+            std::cout << "Running script from " << *args->script << "...";
+        scm_c_primitive_load (args->script->c_str());
+        if (args->verbose)
+            std::cout << "success!" << std::endl;
+    }
+    if (args->interactive)
+    {
+        if (args->verbose)
+            std::cout << "Starting CLI... " << std::endl;
+        scm_c_eval_string ("(start-repl)");
+    }
+    if (gnc_current_session_exist())
+    {
+        if (args->verbose)
+            std::cout << "Warning: session still open. Clearing session..." << std::endl;
+        gnc_clear_current_session ();
+    }
+}
+
+static void load_file (std::string m_file_to_load, bool m_open_readwrite, bool m_verbose)
+{
+    if (m_verbose)
+        std::cout << "\n\nLoading " << m_file_to_load
+                  << (m_open_readwrite ? " (r/w)" : " (readonly)");
+    auto session = gnc_get_current_session();
+    auto mode = m_open_readwrite ? SESSION_READ_ONLY : SESSION_NORMAL_OPEN;
+    while (true)
+    {
+        qof_session_begin (session, m_file_to_load.c_str(), mode);
+        auto io_err = qof_session_get_error (session);
+        switch (io_err)
+        {
+        case ERR_BACKEND_NO_ERR:
+            qof_session_load (session, NULL);
+            if (m_verbose)
+                std::cout << "... done!\n\n";
+            return;
+        case ERR_BACKEND_LOCKED:
+        case ERR_BACKEND_READONLY:
+            if (m_verbose)
+                std::cout << " (forced readonly)";
+            mode = SESSION_READ_ONLY;
+            break;
+        default:
+            if (m_verbose)
+                std::cout << "... unknown error. Abort." << std::endl;
+            exit (1);
+        }
+    }
+}
+
 int
 Gnucash::GnucashCli::start ([[maybe_unused]] int argc, [[maybe_unused]] char **argv)
 {
     Gnucash::CoreApp::start();
 
+    if (m_interactive || m_script)
+    {
```

Comment:
My intention for `Gnucash::GnucashCli::start` was to make it a pure dispatcher for commands provided via the command line prompt. Any actual execution was supposed to go into individual functions in `gnucash-commands.cpp`.

So I would propose to move most of this logic to `gnucash-commands.cpp`.

Note that file loading is used by many commands, so perhaps it's useful to implement your load_file function in there in such a way it can be shared by several other commands.

Reply:
done, moved to `gnucash-commands.cpp`, and shared `load_file` with the cli report code.

Reply:
btw during move into `gnucash-commands.cpp` I saw several variants of shutdown code: `gnc_shutdown_cli`, `scm_cleanup_and_exit_with_failure` etc. a bit confused. during merging `load_file` I also notice the existing code handles a locked file differently; my approch offers to unlock file. This may not be desirable.